x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/TransportPutSamlServiceProviderAction.java

```diff
@@ -0,0 +1,110 @@
+import org.elasticsearch.common.util.iterable.Iterables;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.idp.saml.sp.SamlServiceProviderDocument;
+import org.elasticsearch.xpack.idp.saml.sp.SamlServiceProviderIndex;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Clock;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.stream.Collectors;
+
+public class TransportPutSamlServiceProviderAction
+    extends HandledTransportAction<PutSamlServiceProviderRequest, PutSamlServiceProviderResponse> {
+
+    private final Logger logger = LogManager.getLogger();
+    private final SamlServiceProviderIndex index;
+    private final Clock clock;
+
+    @Inject
+    public TransportPutSamlServiceProviderAction(TransportService transportService, ActionFilters actionFilters,
+                                                 SamlServiceProviderIndex index) {
+        this(transportService, actionFilters, index, Clock.systemUTC());
+    }
+
+    TransportPutSamlServiceProviderAction(TransportService transportService, ActionFilters actionFilters,
+                                                 SamlServiceProviderIndex index, Clock clock) {
+        super(PutSamlServiceProviderAction.NAME, transportService, actionFilters, PutSamlServiceProviderRequest::new);
+        this.index = index;
+        this.clock = clock;
+    }
+
+    @Override
+    protected void doExecute(Task task, final PutSamlServiceProviderRequest request,
+                             final ActionListener<PutSamlServiceProviderResponse> listener) {
+        final SamlServiceProviderDocument document = request.getDocument();
+        if (document.docId != null) {
+            listener.onFailure(new IllegalArgumentException("request document must not have an id [" + document.docId + "]"));
+            return;
+        }
+        index.findByEntityId(document.entityId, ActionListener.wrap(matchingDocuments -> {
+            if (matchingDocuments.isEmpty()) {
+                // derive a document id from the entity id so that don't accidentally create duplicate entities due to a race condition
+                document.docId = deriveDocumentId(document);
+                // force a create in case there are concurrent requests
+                writeDocument(document, DocWriteRequest.OpType.CREATE, listener);
+            } else if (matchingDocuments.size() == 1) {
+                SamlServiceProviderDocument existingDoc = Iterables.get(matchingDocuments, 0);
+                assert existingDoc.docId != null : "Loaded document with no doc id";
+                assert existingDoc.entityId.equals(document.entityId) : "Loaded document with non-matching entity-id";
+                document.setDocId(existingDoc.docId);
+                document.setCreated(existingDoc.created);
+                writeDocument(document, DocWriteRequest.OpType.INDEX, listener);
+            } else {
```

Comment:
Question : Why is this a check we fail specifically upon as opposed to just an assertion like the null Id or entityIds not matching ? Isn't this equally improbable ?

Reply:
No, I don't think it's equally improbable.
The other `assert` checks are supposed to be guaranteed by the index contract. You can't get a doc without an ID from an index, it's impossible unless either:
1. The SamlServiceProviderIndex is horribly broken (unlikely)
2. It's a test with bogus data (much more likely).

Similarly for the `entityId` since the method returns matching docs, it can only represent a completely broken `Index` class, or badly set up mock data.

However, this case is a more realistic scenario that represents some failure to correctly write documents to the index (I don't see how it would happen, but I'm not perfectly confident that we couldn't have such bugs, now or in the future), or it's the result of manual updates of the index (which someone might do, even though they shouldn't).
What should we do in such a case? Since assertions are off in production, if we rely exclusively on `assert` statements, we have no checks in production, so we could either:
1. Treat it like a single element result, and just operate on the first doc. That's not completely terrible, but it's defintely not good. I'd much rather we fail and make the admin fix the broken data
2. Update all the matching docs, which seems even worse.
3. Do nothing, which is clearly much worse.

Reply:
Thanks for going through your reasoning, makes sense to me too